### ISSUE TITLE
HIMloco Sim2Sim And Sim2Real is something wrong

### ISSUE THREAD

**Issue by hulehulema** [fan-ziqi/rl_sar#161]
### Description

The Himloco Sim2Sim and Sim2Real can't works well, because the deploy code with bugs in it.
You can refer to the down code.

https://github.com/Project-Robo/HIMLOCO_GO2

### Commit

main

### Operating System (OS)

22.04

### ROS Version

humble

### Python Version

conda python3.8

**Comment by hulehulema** [comment]
std::vector<float> TorchModel::forward(const std::vector<std::vector<float>>& inputs)
{
    if (!loaded_)
    {
        throw std::runtime_error("Model not loaded");
    }

#ifdef USE_TORCH
    try
    {
        // Convert input vector to Torch tensor (use first input only)
        const auto& input = inputs[0];
        auto input_tensor = torch::tensor(input, torch::kFloat32).reshape({1, static_cast<int64_t>(input.size())});

        // Disable gradient computation before each forward pass
        torch::autograd::GradMode::set_enabled(false);

        // Ensure single-threaded execution (critical for performance!)
        torch::set_num_threads(1);

        // Execute forward inference
        auto output = model_.forward({input_tensor}).toTensor();

        // Convert output tensor to vector
        return torch_to_vector(output);
    }
    catch (const std::exception& e)
    {
        std::cout << LOGGER::ERROR << "Torch inference error: " << e.what() << std::endl;
        throw;
    }
#else
    throw std::runtime_error("Torch support not compiled");
#endif
}
The above code maybe some problem in it.You can't just load the policy and input the obs.There is a latent space,the input should change manual.You can test the code below.

std::vector<float> TorchModel::forward(const std::vector<std::vector<float>>& inputs)
{
    if (!loaded_)
    {
        throw std::runtime_error("Model not loaded");
    }

#ifdef USE_TORCH
    try
    {
        // Convert input vector to Torch tensor (use first input only)
        const auto& input = inputs[0];
        auto input_tensor = torch::tensor(input, torch::kFloat32).reshape({1, static_cast<int64_t>(input.size())});

        // Disable gradient computation before each forward pass
        torch::autograd::GradMode::set_enabled(false);

        // Ensure single-threaded execution (critical for performance!)
        torch::set_num_threads(1);

        //下面为新代码
        //获取子模块
        torch::jit::script::Module actor = model_.attr("actor").toModule();
        torch::jit::script::Module encoder = model_.attr("estimator").toModule();
        
        // 1. 通过encoder得到predict_tensor
        torch::Tensor predict_tensor = encoder.forward({input_tensor}).toTensor();
        
        // 2. 分割为vel_tensor和latent_tensor
        torch::Tensor vel_tensor = predict_tensor.slice(1, 0, 3);
        torch::Tensor latent_tensor = predict_tensor.slice(1, 3);
        
        // 3. 对latent_tensor进行L2归一化
        torch::Tensor norm = torch::norm(latent_tensor, 2, 1, true);
        latent_tensor = latent_tensor / norm;
        
        // 4. 准备obs_tensor输入
        int num_obs_current = 45; // 当前观测维度
        torch::Tensor obs_current = input_tensor.slice(1, 0, num_obs_current);
        torch::Tensor obs_tensor = torch::cat({obs_current, vel_tensor, latent_tensor}, 1);
        
        // 5. 通过actor得到action
        torch::Tensor action_tensor = actor.forward({obs_tensor}).toTensor();
        
        // 6. 转换为std::vector<float>
        action_tensor = action_tensor.detach().cpu().flatten();
        std::vector<float> action;
        
        float* data_ptr = action_tensor.data_ptr<float>();
        long num_elements = action_tensor.numel();
        action.insert(action.end(), data_ptr, data_ptr + num_elements);
        
        return action;
    }
    catch (const std::exception& e)
    {
        std::cout << LOGGER::ERROR << "Torch inference error: " << e.what() << std::endl;
        throw;
    }
#else
    throw std::runtime_error("Torch support not compiled");
#endif
}
